File: packages/core/src/server/routes/automation-webhook.ts
// packages/core/src/server/routes/automation-webhook.ts
//
// Task 25. POST /api/automation-webhooks/:hookId. Auth-exempt by path
// (server/middleware/auth.ts). Owns hookId lookup, header selection, and
// 404/401/204 responses per triggers/webhook.ts's own comment — that module
// stays pure/DB-primitive. Requires the RAW request body: http.ts must mount
// a path-scoped express.raw() ahead of the global express.json() (contract
// §4 — HMAC is computed over exact bytes, not a re-serialized JSON.parse).
import { Router } from 'express';
import type { AutomationDefinition } from '@qlan-ro/mainframe-types';
import type { RouteContext } from './types.js';
import { param } from './types.js';
import { fail, okEmpty } from './respond.js';
import { asyncHandler } from './async-handler.js';
import { createChildLogger } from '../../logger.js';
import { isDedupConflict } from '../../automations/service-helpers.js';
import {
  verifySignature,
  matchPreset,
  deliveryId,
  deliveryTimestampMs,
  isStaleDelivery,
  captureSample,
  GITHUB_PR_OPENED_PRESET,
  GITHUB_PR_MERGED_PRESET,
  type WebhookPresetPredicate,
} from '../../automations/triggers/webhook.js';

const logger = createChildLogger('routes:automation-webhook');

const PRESET_BY_NAME: Record<string, WebhookPresetPredicate> = {
  github_pr_opened: GITHUB_PR_OPENED_PRESET,
  github_pr_merged: GITHUB_PR_MERGED_PRESET,
};

function firstHeader(value: string | string[] | undefined): string | undefined {
  return Array.isArray(value) ? value[0] : value;
}

export function automationWebhookRoutes(ctx: RouteContext): Router {
  const router = Router();

  router.post(
    '/api/automation-webhooks/:hookId',
    asyncHandler(async (req, res) => {
      const service = ctx.automations;
      if (!service) return void fail(res, 503, 'automation service not available');
      const hookId = param(req, 'hookId');

      const found = service.automations.findWebhookTrigger(hookId);
      if (!found) return void fail(res, 404, 'unknown webhook');

      const rawBody = req.body as Buffer;
      const sigHeader = firstHeader(req.headers['x-hub-signature-256']) ?? firstHeader(req.headers['x-signature']);
      const secret = service.credentials.get(`webhook:${hookId}`);
      if (!secret || !verifySignature(secret.token, rawBody, sigHeader)) {
        return void fail(res, 401, 'invalid signature');
      }

      let payload: Record<string, unknown>;
      try {
        payload = JSON.parse(rawBody.toString('utf8')) as Record<string, unknown>;
      } catch {
        return void fail(res, 400, 'invalid JSON payload');
      }
      const githubEvent = firstHeader(req.headers['x-github-event']);
      if (githubEvent) payload.event = githubEvent;

      const preset = found.trigger.preset ? PRESET_BY_NAME[found.trigger.preset] : undefined;
      if (preset && !matchPreset(preset, payload)) {
        res.status(204).end();
        return;
      }

      const timestampMs = deliveryTimestampMs(payload, req.headers as Record<string, string | string[] | undefined>);
      if (timestampMs !== null && isStaleDelivery(timestampMs, Date.now())) {
        logger.warn({ hookId, timestampMs }, 'stale webhook delivery dropped');
        res.status(204).end();
        return;
      }

      let delivery: string;
      try {
        delivery = deliveryId(payload, req.headers as Record<string, string | string[] | undefined>);
      } catch (err) {
        return void fail(res, 400, err instanceof Error ? err.message : String(err));
      }

      captureSample(service.db, found.row.id, found.trigger.id, payload);

      // Disabled automations stay a silent 200 (matches TriggerArmer.fireRun's
      // own enabled check) so a disabled automation's webhook doesn't leak its
      // existence via a differing status — findWebhookTrigger deliberately
      // includes disabled rows for the same reason.
      if (found.row.enabled === 1) {
        // Bypasses TriggerArmer.fireRun (which swallows every startRun error,
        // dedup or not, since its other callers have no HTTP response to fail)
        // so this route can tell the two apart: a duplicate delivery is a
        // no-op 200, any other failure to start the run is a 500 the sender
        // should retry (contract A7).
        try {
          const definition = JSON.parse(found.row.definition) as AutomationDefinition;
          const run = service.interpreter.startRun(
            found.row.id,
            definition,
            { kind: 'webhook', triggerId: found.trigger.id, payload },
            `${found.trigger.id}|${delivery}`,
          );
          void service.interpreter.advance(run.id).catch((err: unknown) => {
            logger.error({ err, runId: run.id }, 'webhook delivery: advance failed');
          });
        } catch (err) {
          if (!isDedupConflict(err)) {
            logger.error({ err, hookId }, 'webhook delivery failed to start a run');
            return void fail(res, 500, 'failed to start automation run');
          }
        }
      }
      okEmpty(res);
    }),
  );

  return router;
}

File: packages/core-rs/crates/mainframe-server/src/routes/automation_webhook.rs
//! `POST /api/automation-webhooks/:hookId` — auth-exempt by path
//! (middleware/auth.rs). HMAC is computed over the exact request bytes, so
//! the handler takes `Bytes`, never a parsed JSON extractor. Status mapping
//! is A7: duplicate 200 no-op, preset non-match / stale 204, start failure
//! 500 (the sender retries).

use std::sync::Arc;

use axum::Router;
use axum::body::Bytes;
use axum::extract::{Path, State};
use axum::http::{HeaderMap, StatusCode};
use axum::response::{IntoResponse, Response};
use axum::routing::post;
use mainframe_automations::triggers::{WebhookDecision, WebhookHeaders};

use crate::ctx::AppCtx;
use crate::respond::{fail, ok_empty};
use crate::routes::automations::{engine, unavailable};

fn header(headers: &HeaderMap, name: &str) -> Option<String> {
    headers
        .get(name)
        .and_then(|value| value.to_str().ok())
        .map(str::to_string)
}

async fn ingest(
    State(ctx): State<Arc<AppCtx>>,
    Path(hook_id): Path<String>,
    headers: HeaderMap,
    body: Bytes,
) -> Response {
    let Some(engine) = engine(&ctx) else {
        return unavailable();
    };
    let webhook_headers = WebhookHeaders {
        signature: header(&headers, "x-hub-signature-256")
            .or_else(|| header(&headers, "x-signature")),
        github_event: header(&headers, "x-github-event"),
        github_delivery: header(&headers, "x-github-delivery"),
        timestamp: header(&headers, "x-timestamp"),
    };
    match engine
        .process_webhook(&hook_id, &webhook_headers, &body)
        .await
    {
        WebhookDecision::UnknownHook => fail(StatusCode::NOT_FOUND, "unknown webhook"),
        WebhookDecision::InvalidSignature => fail(StatusCode::UNAUTHORIZED, "invalid signature"),
        WebhookDecision::InvalidJson => fail(StatusCode::BAD_REQUEST, "invalid JSON payload"),
        WebhookDecision::MissingDeliveryId => fail(
            StatusCode::BAD_REQUEST,
            "webhook delivery missing a delivery id (X-GitHub-Delivery header or payload.id)",
        ),
        WebhookDecision::PresetMismatch | WebhookDecision::StaleDelivery => {
            StatusCode::NO_CONTENT.into_response()
        }
        WebhookDecision::Duplicate | WebhookDecision::Accepted { .. } => ok_empty(),
        WebhookDecision::StartFailed { error } => {
            tracing::error!(hook_id, error, "webhook delivery failed to start a run");
            fail(
                StatusCode::INTERNAL_SERVER_ERROR,
                "failed to start automation run",
            )
        }
    }
}

pub fn router() -> Router<Arc<AppCtx>> {
    Router::new().route("/api/automation-webhooks/{hookId}", post(ingest))
}

#[cfg(test)]
mod automation_webhook_tests;

// PORT STATUS: src/server/routes/automation-webhook.ts (124 lines)
// confidence: high
// todos: 0
// notes: decision→status table mirrors triggers/webhook_ingest.rs's module
//        doc exactly; the ingest pipeline itself is engine-side (T8.3).
